subsys/sd/sdio.c

```diff
@@ -56,23 +63,995 @@ int sdio_send_ocr(struct sd_card *card, uint32_t ocr)
+	k_mutex_unlock(&func->card->lock);
+	return ret;
+}
+
+/**
+ * @brief Write blocks to SDIO fifo
+ *
+ * Writes blocks from SDIO register, treating it as a fifo. Writes will
+ * all be done to same address.
+ * @param func: function to write to
+ * @param reg: register address of fifo
+ * @param data: data to write to fifo
+ * @param blocks: number of blocks to write to fifo
+ * @retval 0 write succeeded
+ * @retval -EBUSY: card is busy with another request
+ * @retval -ETIMEDOUT: card write timed out
+ * @retval -EIO: I/O error
+ */
+int sdio_write_blocks_fifo(struct sdio_func *func, uint32_t reg, uint8_t *data,
+	uint32_t blocks)
+{
+	int ret;
+
+	if ((func->card->type != CARD_SDIO) && (func->card->type != CARD_COMBO)) {
+		LOG_WRN("Card does not support SDIO commands");
+		return -ENOTSUP;
+	}
+	ret = k_mutex_lock(&func->card->lock, K_NO_WAIT);
+	if (ret) {
+		LOG_WRN("Could not get SD card mutex");
+		return -EBUSY;
+	}
+	ret = sdio_io_rw_extended(func->card, SDIO_IO_WRITE, func->num, reg,
+		false, data, blocks, func->block_size);
+	k_mutex_unlock(&func->card->lock);
+	return ret;
+}
+
+/**
+ * @brief Copy bytes from an SDIO card
+ *
+ * Copies bytes from an SDIO card, starting from provided address.
+ * @param func: function to read from
+ * @param reg: register address to start copy at
+ * @param data: buffer to copy data into
+ * @param len: length of data to read
+ * @retval 0 read succeeded
+ * @retval -EBUSY: card is busy with another request
+ * @retval -ETIMEDOUT: card read timed out
+ * @retval -EIO: I/O error
+ */
+int sdio_read_addr(struct sdio_func *func, uint32_t reg, uint8_t *data,
+	uint32_t len)
+{
+	int ret;
+
+	if ((func->card->type != CARD_SDIO) && (func->card->type != CARD_COMBO)) {
+		LOG_WRN("Card does not support SDIO commands");
+		return -ENOTSUP;
+	}
+	ret = k_mutex_lock(&func->card->lock, K_NO_WAIT);
+	if (ret) {
+		LOG_WRN("Could not get SD card mutex");
+		return -EBUSY;
+	}
+	ret = sdio_io_rw_extended_helper(func, SDIO_IO_READ, reg, true,
+		data, len);
+	k_mutex_unlock(&func->card->lock);
+	return ret;
+}
+
+/**
+ * @brief Copy bytes to an SDIO card
+ *
+ * Copies bytes to an SDIO card, starting from provided address.
+ *
+ * @param func: function to write to
+ * @param reg: register address to start copy at
+ * @param data: buffer to copy data from
+ * @param len: length of data to write
+ * @retval 0 write succeeded
+ * @retval -EBUSY: card is busy with another request
+ * @retval -ETIMEDOUT: card write timed out
+ * @retval -EIO: I/O error
+ */
+int sdio_write_addr(struct sdio_func *func, uint32_t reg, uint8_t *data,
+	uint32_t len)
+{
+	int ret;
+
+	if ((func->card->type != CARD_SDIO) && (func->card->type != CARD_COMBO)) {
+		LOG_WRN("Card does not support SDIO commands");
+		return -ENOTSUP;
+	}
+	ret = k_mutex_lock(&func->card->lock, K_NO_WAIT);
+	if (ret) {
+		LOG_WRN("Could not get SD card mutex");
+		return -EBUSY;
+	}
+	ret = sdio_io_rw_extended_helper(func, SDIO_IO_WRITE, reg, true,
+		data, len);
```

Comment:
I think you should do a pass with the right vim indentation settings to align to the previous open parenthesis and fix all of these, they just look weird, to me at least :-) apart from that good stuff!

Reply:
I think I've caught the remainder of them in latest push- let me know if you see any others